### PR TITLE
Generate requirements files from pyproject.toml

### PR DESCRIPTION
**Pull request by mwtoews**
This PR is based on https://github.com/numpy/numpydoc/pull/483 which makes `pyproject.toml` a primary file without any template, and adds a pre-commit script to write requirements files. Some of the comments from the requirements files are moved to `pyproject.toml`.

### CODE REVIEW

**Comment by mwtoews** [comment]
Question: why is there a "default" extra? It repeats the same info in the dependencies. The script can easily be modified to generate this file too, if that extra were removed.

**Comment by lagru** [comment]
Hey @mwtoews thanks for the suggestion! It seems useful to me to have the source of truth all originate from one place. Especially because it may make it easier to use comments in the pyproject.toml. Before I review, I am curious, did the previous discussion in numpydoc bring up any other pros or cons?

The `default.txt` is used in several places in the CI to install the runtime dependencies in a controlled manner independent of the package itself e.g. see [this search](https://github.com/search?q=repo%3Ascikit-image%2Fscikit-image%20default.txt&type=code). I am not sure if there is a clear advantage to doing it that way but for now would hope to keep the noise to our infrastructure minimal.

File: tools/generate_requirements.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+try:  # standard module since Python 3.11
+    import tomllib as toml
+except ImportError:
+    try:  # available for older Python via pip
+        import tomli as toml
+    except ImportError:
+        sys.exit("Please install `tomli` first: `pip install tomli`")
+
+repo_dir = (Path(__file__).parent / "..").resolve()
```
Comment (lagru) [inline tools/generate_requirements.py:15]:
```suggestion
repo_dir = Path(__file__).parent.parent
```
I think that gives the same result?

File: tools/generate_requirements.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+req_dir = repo_dir / "requirements"
+pyproject = toml.loads((repo_dir / "pyproject.toml").read_text())
```
Comment (lagru) [inline tools/generate_requirements.py:17]: Huh, didn't know about [`Path.read_text()`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.Path.read_text). Nice.

File: requirements/developer.txt [hunk 3]
```diff
@@ -1,2 +1,3 @@
+# Generated from pyproject.toml
 pre-commit
-rtoml
+tomli; python_version < '3.11'
```
[the remark this answers is not in the document]
Comment (mwtoews) [reply requirements/developer.txt:3]: Yes, I found that omission with conda-forge too.

From what I understand, tomli is the backport library for tomllib with Python 3.11+, so it should be the best forward-looking option until py311 is the minimum version in a couple of years.

File: tools/generate_requirements.py (hunk 1 above)
Comment (mwtoews) [reply tools/generate_requirements.py]: Yes it does, thanks!

**Comment by mwtoews** [comment]
> Before I review, I am curious, did the previous discussion in numpydoc bring up any other pros or cons?

Most of the discussion was from a superseded PR https://github.com/numpy/numpydoc/pull/481 where I was aiming to use a dynamic tool-dependant feature of setuptools to read optional dependencies (rather than using a script and template approach). Ultimately, there are more strengths in having a single statically-analyzable pyproject.toml (complete with all dependencies), but some tools also need to `pip install -r` with individual requirements files too. The applied PR for numpydoc and in this PR is to also remove the template approach, since requirements files don't really need a template (except for a comment header). A clear benefit is that edits to pyproject.toml now directly go to that file, and where present, editors may enable syntax highlighting (template files can't do that).

As for the requirements/default.txt file, I have not problems with it. But in pyproject.toml it is duplicated. From pip's perspective `pip install scikit-image` and `pip install scikit-image[default]` do the same thing. My guess is removing the `default` extra and generating a requirements files from `[project.dependencies]` will not change anything, and resolve the duplication.

File: tools/generate_requirements.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    try:  # available for older Python via pip
+        import tomli as toml
+    except ImportError:
+        sys.exit("Please install `tomli` first: `pip install tomli`")
+
+repo_dir = Path(__file__).parent.parent
+
+
+def generate_requirement_file(name: str, req_list: list[str]) -> None:
+    lines = ["# Generated from pyproject.toml, do not edit"] + req_list
```
Comment (lagru) [inline tools/generate_requirements.py:19]: Sorry, maybe I should have been clearer in https://github.com/scikit-image/scikit-image/pull/7085#discussion_r1284413772. I meant to include the origin of the automatically generated file. E.g.

```python
# Generated from pyproject.toml via tools/generate_requirements.py and pre-commit hook
# do not edit manually
```

File: pyproject.toml [hunk 5]
```diff
@@ -63,24 +54,14 @@ build = [
     'Cython>=0.29.32',
     'pythran',
     'numpy>=1.22',
+    # Developer UI
     'spin==0.4',
     'build',
 ]
 data = ['pooch>=1.6.0']
-default = [
```
Comment (lagru) [inline pyproject.toml:70]: > [...] But in pyproject.toml it is duplicated. From pip's perspective pip install scikit-image and pip install scikit-image[default] do the same thing. [...] -- https://github.com/scikit-image/scikit-image/pull/7085#issuecomment-1666142890

Ah, got you, nice catch. Happy to remove it. :)

**Review by lagru — COMMENTED** [review]
Thanks for addressing my suggestions! I have only a minor one left.

File: tools/generate_requirements.py (hunk 4 above)
Comment (mwtoews) [reply tools/generate_requirements.py]: I've added a few more helpful details to the header.

File: pyproject.toml (hunk 5 above)
Comment (mwtoews) [reply pyproject.toml:70]: If, by chance, anyone tries `pip install --dry-run scikit-image[default]` they would see "WARNING: scikit-image 0.22.0rc0 does not provide the extra 'default'". But the installed set of packages is the same.

Comment (lagru) [reply pyproject.toml:70]: Perfect. :)

**Pull request merged**